drivers/modem/simcom-sim7080.c

```diff
@@ -0,0 +1,2376 @@
+
+static struct k_thread modem_rx_thread;
+static struct k_work_q modem_workq;
+static struct sim7080_data mdata;
+static struct modem_context mctx;
+static const struct socket_op_vtable offload_socket_fd_op_vtable;
+
+static struct zsock_addrinfo dns_result;
+static struct sockaddr dns_result_addr;
+static char dns_result_canonname[DNS_MAX_NAME_SIZE + 1];
+
+static struct sim7080_gnss_data gnss_data;
+
+static K_KERNEL_STACK_DEFINE(modem_rx_stack, CONFIG_MODEM_SIMCOM_SIM7080_RX_STACK_SIZE);
+static K_KERNEL_STACK_DEFINE(modem_workq_stack, CONFIG_MODEM_SIMCOM_SIM7080_RX_WORKQ_STACK_SIZE);
+NET_BUF_POOL_DEFINE(mdm_recv_pool, MDM_RECV_MAX_BUF, MDM_RECV_BUF_SIZE, 0, NULL);
+
+static void socket_close(struct modem_socket *sock);
+const struct socket_dns_offload offload_dns_ops;
+
+static inline uint32_t hash32(char *str, int len)
+{
+#define HASH_MULTIPLIER 37
+	uint32_t h = 0;
+	int i;
+
+	for (i = 0; i < len; ++i) {
+		h = (h * HASH_MULTIPLIER) + str[i];
+	}
+
+	return h;
+}
+
+static inline uint8_t *modem_get_mac(const struct device *dev)
+{
+	struct sim7080_data *data = dev->data;
+	uint32_t hash_value;
+
+	data->mac_addr[0] = 0x00;
+	data->mac_addr[1] = 0x10;
+
+	/* use IMEI for mac_addr */
+	hash_value = hash32(mdata.mdm_imei, strlen(mdata.mdm_imei));
+
+	UNALIGNED_PUT(hash_value, (uint32_t *)(data->mac_addr + 2));
+
+	return data->mac_addr;
+}
+
+/* Setup the Modem NET Interface. */
+static void modem_net_iface_init(struct net_if *iface)
+{
+	const struct device *dev = net_if_get_device(iface);
+	struct sim7080_data *data = dev->data;
+
+	net_if_set_link_addr(iface, modem_get_mac(dev), sizeof(data->mac_addr), NET_LINK_ETHERNET);
+
+	data->netif = iface;
+
+	socket_offload_dns_register(&offload_dns_ops);
+}
+
+/**
+ * Changes the operating state of the sim7080.
+ *
+ * @param state The new state.
+ */
+static void change_state(enum sim7080_state state)
+{
+	LOG_DBG("Changing state to (%d)", state);
+	mdata.state = state;
+}
+
+/**
+ * Get the current operating state of the sim7080.
+ *
+ * @return The current state.
+ */
+static enum sim7080_state get_state(void)
+{
+	return mdata.state;
+}
+
+/*
+ * Parses the +CAOPEN command and gives back the
+ * connect semaphore.
+ */
+MODEM_CMD_DEFINE(on_cmd_caopen)
+{
+	int result = atoi(argv[1]);
+
+	LOG_INF("+CAOPEN: %d", result);
+	modem_cmd_handler_set_error(data, result);
+	return 0;
+}
+
+/*
+ * Unlock the tx ready semaphore if '> ' is received.
+ */
+MODEM_CMD_DIRECT_DEFINE(on_cmd_tx_ready)
+{
+	k_sem_give(&mdata.sem_tx_ready);
+	return len;
+}
+
+/*
+ * Connects an modem socket. Protocol can either be TCP or UDP.
+ */
+static int offload_connect(void *obj, const struct sockaddr *addr, socklen_t addrlen)
+{
+	struct modem_socket *sock = (struct modem_socket *)obj;
+	uint16_t dst_port = 0;
+	char *protocol;
+	struct modem_cmd cmd[] = { MODEM_CMD("+CAOPEN: ", on_cmd_caopen, 2U, ",") };
+	char buf[sizeof("AT+CAOPEN: #,#,#####,####.###.###.####,####")];
+	char ip_str[NET_IPV4_ADDR_LEN];
```

Comment:
Do you only support IPv4? This looks like it may not work for IPv6.

Reply:
Yes, currently only ipv4 is supported

Reply:
Expanded the buffer to support ipv6